src/ApiService/ApiService/onefuzzlib/JobResultOperations.cs

```diff
@@ -2,99 +2,69 @@
+        return await data.FirstOrDefaultAsync();
     }
 
-    private JobResult UpdateResult(JobResult result, JobResultType type, Dictionary<string, double> resultValue) {
+    private async Async.Task<bool> TryUpdate(Job job, Guid taskId, Guid machineId, string resultType, Dictionary<string, double> resultValue) {
+        var jobId = job.JobId;
+        var machineIdMetric = string.Concat(machineId, resultType);
 
-        var newResult = result;
-        double newValue;
-        switch (type) {
-            case JobResultType.NewCrashingInput:
-                newValue = result.NewCrashingInput + resultValue["count"];
-                newResult = result with { NewCrashingInput = newValue };
-                break;
-            case JobResultType.NewReport:
-                newValue = result.NewReport + resultValue["count"];
-                newResult = result with { NewReport = newValue };
-                break;
-            case JobResultType.NewUniqueReport:
-                newValue = result.NewUniqueReport + resultValue["count"];
-                newResult = result with { NewUniqueReport = newValue };
-                break;
-            case JobResultType.NewRegressionReport:
-                newValue = result.NewRegressionReport + resultValue["count"];
-                newResult = result with { NewRegressionReport = newValue };
-                break;
-            case JobResultType.NewCrashDump:
-                newValue = result.NewCrashDump + resultValue["count"];
-                newResult = result with { NewCrashDump = newValue };
-                break;
-            case JobResultType.CoverageData:
-                double newCovered = resultValue["covered"];
-                double newTotalCovered = resultValue["features"];
-                double newCoverageRate = resultValue["rate"];
-                newResult = result with { InstructionsCovered = newCovered, TotalInstructions = newTotalCovered, CoverageRate = newCoverageRate };
+        var oldEntry = await GetJobResult(taskId, machineIdMetric);
+        var newEntry = new JobResult(TaskId: taskId, MachineIdMetric: machineIdMetric, JobId: jobId, Project: job.Config.Project, Name: job.Config.Name, resultType, resultValue);
+
+        if (oldEntry == null) {
+            _logTracer.LogInformation($"attempt to insert new job result {taskId} and machineId+metricType {machineIdMetric}");
+            var result = await Insert(newEntry);
+            if (!result.IsOk) {
+                throw new InvalidOperationException($"failed to insert job result with taskId {taskId} and machineId+metricType {machineIdMetric}");
+            }
+            return true;
+        }
+
+        ResultVoid<(HttpStatusCode Status, string Reason)> r;
+        switch (resultType) {
+            case COVERAGE_DATA when oldEntry.MetricValue["rate"] < newEntry.MetricValue["rate"]:
+                r = await Replace(newEntry);
```

Comment:
we should use update instead of replace

Reply:
I purposefully used replace over update because it significantly decreases the collisions inherent to updating the table. Because we only actually need the max, most recent value from the metric (iteration count, coverage rate, etc.), we can overwrite the existing entry without having to worry about updating in the right order.

Reply:
We _shouldn't_ get collisions here with Update since we're expecting this value to be updated in a "single-threaded" manner with only one writer. But it protects us in the case that there has been a problem and two updates come in at once:

1. "Thread" A gets entity, rate = 0
2. "Thread" B gets entity, rate = 0 
3. B compares with rate 2, it's bigger, so writes entity
4. A compares with rate 1, it's bigger, so writes entity - rate has been decreased from 2 to 1

----

Alternately, if we really don't care about checking for collisions on this entity (we can make a strong assumption that the updates are "single-threaded"), then we can do a simple Replace without fetching the entity first, since we don't actually need to compare with the existing value. This would half the number of table operations performed.

Reply:
There can be multiple instances of the same function running at the same time, so I don't think we can treat the function as one writer. Your example makes a good case for Update, though. I was just trying to reduce collisions, which definitely seem to be happening - it's why I initially wrote the `polly` policy. 

The latter scenarios would not be possible, though. It's entirely possible for files that get uploaded to the queue by the agent get out of order, no? Furthermore, with multiple function workers, items can get out of order as they are de-queued.

Reply:
The "one writer" that I'm thinking is per job/task/machine/resultType. But yes if this isn't coming from the agent itself but instead from one of the queues then we can't assume that.

Reply:
It's impossible to avoid collisions when writing to the table unless we use a lock. I don't really understand why ordering is an issue here, but to solve it you have 2 options (assuming we're also using a lock):

1. Include a UTC timestamp in the message and only update if the timestamp is newer. We know a collision in a row of the JobResult table can only happen if the messages come from the same machine. Since the messages come from the same machine, the timestamps from the messages come from the same clock and we can assume the clock is internally consistent so we don't need to worry about any clock sync issues.
2. Don't store order sensitive data. Instead of storing a 'rate', and needing to keep track of the latest value; store 'iteration count' which only requires incrementing the value (which is safe under a lock) and a start time

Reply:
@tevoinea, ordering is an issue for both the `CoverageData` and the `RuntimeStats` metrics. `CoverageData` has three metrics that we want to track (rate, covered, & features). `RuntimeStats` has iteration_count. 

All of these metric fields require we take the most recent max. Iteration count sends the total iterations up to time of emission, so your point about 2 does not hold. This is also why ordering matters. If a bunch of `RuntimeStats` end up in the queue out of order, then we need to take only take the max value up to that point. 

Wrt to your point on 1., this had already been suggested, but I considered the value comparison (only update when newRuntimeStats > oldRuntimeStats) to be equivalent. If we don't think this is sufficient then I can just include the timestamp and run with that.

Reply:
@nharper285 makes sense, the benefit I see with using timestamp is that we can use the same comparison method for all event types meaning we don't need to check `rate` for coverage, `total_count` for runtime_stats, etc. Also, it will still work if we ever produce event types that don't contain ordinal data.

Reply:
That's a good point! I'll make the update.

Reply:
@tevoinea, let me know when you get a chance to review updates.

Reply:
> I don't really understand why ordering is an issue here

It's an issue because these updates are apparently driven by queue messages. We don't know how long they've been on the queue or how many other processors of queue messages there are. 

If this was an HTTP API that was being directly invoked from the agent machine then we could simply do the update without checking (this is the "one writer" scenario).